include/deal.II/base/hdf5.h

```diff
@@ -0,0 +1,712 @@
+ * to collective in order to maximize the performance. This means that all the
+ * MPI processes have to contribute to every single call, even if they don't
+ * have data to write.
+ *
+ * ## Write a hyperslab in parallel
+ * The example below shows how to write a hyperslab.
+ * @code
+ * std::vector<hsize_t> dataset_dimensions = {50, 30};
+ * auto dataset = group.create_dataset<double>("name", dataset_dimensions);
+ *  if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+ *    {
+ *      // data can be std::vector, FullMatrix or Vector
+ *      FullMatrix<double> data = {...};
+ *      std::vector<hsize_t> hyperslab_dimensions = {2, 5};
+ *      std::vector<hsize_t> hyperslab_offset     = {0, 0};
+ *      dataset.write_hyperslab(hyperslab_data,
+ *                              hyperslab_offset,
+ *                              hyperslab_dimensions);
+ *    }
+ *  else
+ *    {
+ *      dataset.write_none<double>();
+ *    }
+ * @endcode
+ *
+ * ## Write unordered data in parallel
+ * The example below shows how to write a selection of data.
+ * @code
+ * std::vector<hsize_t> dataset_dimensions = {50, 30};
+ * auto dataset = group.create_dataset<double>("name", dataset_dimensions);
+ * std::vector<hsize_t> coordinates_a = {0,
+ *                                       0, // first point
+ *                                       0,
+ *                                       2, // second point
+ *                                       3,
+ *                                       4, // third point
+ *                                       25,
+ *                                       12}; // fourth point
+ *  std::vector<double>  data_a        = {2, 3, 5, 6};
+ *
+ *  std::vector<hsize_t> coordinates_b = {5,
+ *                                        0, // first point
+ *                                        0,
+ *                                        4, // second point
+ *                                        5,
+ *                                        4, // third point
+ *                                        26,
+ *                                        12}; // fourth point
+ *  std::vector<double>  data_b        = {9, 4, 7, 6};
+ *  if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+ *    {
+ *      dataset.write_selection(data_a, coordinates_a);
+ *    }
+ *  else if (Utilities::MPI::this_mpi_process(mpi_communicator) == 1)
+ *    {
+ *      dataset.write_selection(data_b, coordinates_b);
+ *    }
+ *  else
+ *    {
+ *      dataset.write_none<double>();
+ *    }
+ * @endcode
+ *
+ * ## Query the type of I/O that HDF5 performed on the last parallel I/O call
+ * In some cases such as when there is type conversion the HDF5 library can
+ * decide to do independent I/O instead of collective I/O. The following code
+ * can be used to query the I/O method.
+ * @code
+ * auto dataset = group.create_dataset<double>("name", dimensions);
+ * #ifdef DEBUG
+ * dataset.check_io_mode(true);
+ * #endif
```

Comment:
HDF5 is very flexible and allows to do very fancy operations such as type conversion, compression on the fly, etc. Not all the operations are compatible with collective MPI. The operation will always succeed but it may not be collective, therefore the performance won't be optimum.

As a matter of fact. In one of the tests I do type conversion and request a collective operation. As expected, the data is correctly written but  get_global_no_collective_cause tells that the operation was not collective because of the type conversion.

Mode conversion happens when you create a `float` dataset and store `double` data in the dataset.

I'll add support for compression on the fly with MPI in a future PR.

**I added the following text in the documentation:**

The default access mode in the HDF5 C++ interface of deal.ii is collective which is typically faster since it allows MPI to do more optimizations. In some cases, such as when there is type conversion, the HDF5 library can decide to do independent I/O instead of collective I/O, even if the user asks for collective I/O. See the following [article](https://www.hdfgroup.org/2015/08/parallel-io-with-hdf5/)

In cases where maximum performance is a requirement, it is important to make sure that all MPI read/write operations are collective. The HDF5 library provides API routines that can be used after the read/write I/O operations to query the I/O mode. If DataSet::query_io_mode() is set to True, then after every read/write operation the HDF5 deal.ii interface calls the routines [H5Pget_mpio_actual_io_mode()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioActualIoMode) and [H5Pget_mpio_no_collective_cause()](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetMpioNoCollectiveCause). The results are stored in `DataSet::io_mode`, `DataSet::local_no_collective_cause` and `DataSet::get_global_no_collective_cause`. We suggest to query the I/O mode only in Debug mode because it requires calling additional HDF5 routines.

**I added as well:**
 If the write operation was collective then the output should be
 ````
 IO mode: H5D_MPIO_CONTIGUOUS_COLLECTIVE
 Local no collective cause: H5D_MPIO_COLLECTIVE
 Global no collective cause: H5D_MPIO_COLLECTIVE
 ````
 See DataSet::io_mode(), DataSet::local_no_collective_cause() and
 DataSet::get_global_no_collective_cause() for all the possible return
 codes.

Reply:
Thanks for adding all of this documentation. 

My comment was more about the naming of the function. You may have renamed it in the meantime, but a function that is called `check_xxx()` suggests that it is doing something that returns a value that is of use to the caller. In fact, I would expect a function called `check_XXX()` to be a `const` member function, and it makes no sense to call a `const` member function if you are not interested in the return type -- because a `const` member function does not change the state, and calling it without doing something with the return type is then a no-op :-)